[16.0][ADD] stock_picking_operation_loss_quantity

rousseldenis opened the pull request:
Depends on:

- #1330

rousseldenis committed:
[16.0][ADD] stock_picking_operation_loss_quantity

rousseldenis committed:
[IMP] stock_picking_operation_loss_quantity: Add activity schedule

lmignon committed:
[IMP] stock_picking_operation_loss_quantity: cancel loss reservation on inventory

 When an inventory is validated, we need to cancel any remaining pending moves created to make the quantity no more available in case of loss declaration.

lmignon commented:
depends on #1330

rousseldenis commented:
What this module needs is to make the user access inventory easily through the Loss picking menu

rousseldenis committed:
[IMP] stock_picking_operation_loss_quantity: Be less restrictive to cancel the loss move

When user want to apply inventory, be less restrictive on inventory_mode
check as the values should have been checked in stock module.

rousseldenis committed:
[IMP] stock_picking_loss_quantity: Don't lock void ids

lmignon commented:
ping @rousseldenis This one is used in our project and in production for years...

nicolas-delbovier-acsone committed:
[IMP] stock_picking_operation_loss_quantity: dynamic sequence prefix

Update the warehouse loss sequence creation to use the specific code
defined in the loss picking type. If no code is defined, it falls
back to the default "LOSS" string.

nicolas-delbovier-acsone committed:
[REF] stock_picking_operation_loss_quantity: get rid of the useless wizards + clean code

- Relocate the loss declaration code from external wizard to the stock.move.line model
- Remove useless grouping logic (the code is always called on a single line through UI)
- Rename variables in for better clarity
- Remove code that was never reached in unit tests and that seemed dead
- Clean and improve unit tests (add quants verifications)

nicolas-delbovier-acsone committed:
[IMP] stock_picking_operation_loss_quantity: link similar loss move lines to same loss picking

nicolas-delbovier-acsone committed:
[FIX] stock_picking_operation_loss_quantity: fix pre-commit

nicolas-delbovier-acsone committed:
[FIX] stock_picking_operation_loss_quantity: make sure the loss moves do not get merged

nicolas-delbovier-acsone committed:
[IMP] stock_picking_operation_loss_quantity: auto-clear stock on threshold

Introduce a 'Loss Auto-Clear Threshold' setting on the warehouse.
When a product/location combination (quant) reaches this number of
consecutive loss declarations, the system automatically reserves the remaining quantity.

This prevents redundant manual declarations for "ghost" stock and
improves warehouse efficiency by stopping future workers from
attempting to pick non-existent products.

nicolas-delbovier-acsone committed:
[FIX] stock_picking_operation_loss_quantity: do not allow to declare a loss on loss pickings

nicolas-delbovier-acsone committed:
[FIX] stock_picking_operation_loss_quantity: use moves instead of move lines to count number of loss declaration on loss picking

Each loss declaration creates a single move on the loss picking but it is not sure that every move has a single move line

nicolas-delbovier-acsone committed:
[FIX] stock_picking_operation_loss_quantity: ensure every loss move is confirmed and linked to a move line when using `loss_auto_clear_threshold=1`

nicolas-delbovier-acsone committed:
[IMP] stock_picking_operation_loss_quantity: Add unit test to show that auto-clearing parent location also clears child location

lmignon commented on stock_picking_operation_loss_quantity/models/stock_move_line.py:
```diff
@@ -0,0 +1,173 @@
+            "location_dest_id": loss_pick_type.default_location_dest_id.id,
+            "lot_ids": [Command.set(self.lot_id.ids)] if self.lot_id else False,
+        }
+
```
@nicolas-delbovier-acsone For flexibility and readability this code should be put into a `_prepare_loss_move_vals(self, qty)` method `

lmignon commented on stock_picking_operation_loss_quantity/models/stock_move_line.py:
```diff
@@ -0,0 +1,173 @@
+                loss_pick_type.default_location_dest_id.id,
+            ),
+        ]
+        similar_loss_lines = self.env["stock.move.line"].search(search_domain)
```
IMO This code should also be put into a dedicated method and the domain also provided by a specific method

lmignon commented on stock_picking_operation_loss_quantity/models/stock_move_line.py:
```diff
@@ -0,0 +1,173 @@
+                    }
+                )
+                new_loss_move._action_confirm(merge=False)
+
```
I can see a problem with this approach. In the end, if the product hasn’t been found, we’re left with n identical moves plus a final move with the remaining quantities in the location. For the operator, who only sees the transactions, this will pose a problem. In my view, it would be preferable to merge the moves and retain the number of loss declarations already made for the same product/batch in the same location.

<img width="1319" height="391" alt="Image" src="https://github.com/user-attachments/assets/7df423dd-c59c-4862-aefb-12281b978ba8" />

lmignon commented on stock_picking_operation_loss_quantity/models/stock_move_line.py:
```diff
@@ -0,0 +1,173 @@
+                {
+                    "picking_type_id": loss_pick_type.id,
+                    "location_id": self.location_id.id,
+                    "location_dest_id": loss_pick_type.default_location_dest_id.id,
```
IMO the information required to create the picking should be provided by a prepare method `_prepare_loss_picking_vals` ....

lmignon reviewed (changes requested):
@nicolas-delbovier-acsone Thank you for the improvements. However, I do have some concerns about the usability of the approach. See my comment.

nicolas-delbovier-acsone committed:
[REF] stock_picking_operation_loss_quantity: extract logic into methods

Refactor the module to improve extensibility and readability by extracting
search domains and write/create values into dedicated methods.

nicolas-delbovier-acsone committed:
[REF] stock_picking_operation_loss_quantity: rename method to avoid WMS ambiguity

Rename `_release_unprocessed_qty` to `_unreserve_unprocessed_qty`.

The term 'release' can be confused with WMS-specific
logic. Using 'unreserve' better aligns with standard Odoo stock terminology
and clarifies that the function handles the unreservation of quantities.

nicolas-delbovier-acsone committed:
[IMP] stock_picking_operation_loss_quantity: use a counter for the number of loss declaration instead of relying on number of moves on the picking

nicolas-delbovier-acsone committed:
[FIX] stock_picking_operation_loss_quantity: ensure all moves get merged in loss picking

nicolas-delbovier-acsone committed:
[FIX] stock_picking_operation_loss_quantity: do not use `self` in loop

lmignon commented on stock_picking_operation_loss_quantity/models/stock_move_line.py:
```diff
@@ -0,0 +1,185 @@
+            ),
+        ]
+
+    def _find_loss_picking(self):
```
@nicolas-delbovier-acsone You search for a `stock.move.line` not for a picking. This method should be renamed.

nicolas-delbovier-acsone replied:
No, I am indeed looking for a picking.

I look for a picking of type "LOSS" that contains moves that match with my new one I want to create: 
```python
def _find_loss_picking(self):  
    similar_loss_lines = self.env["stock.move.line"].search(
        self._find_loss_picking_moves_domain()
    )
    loss_picking = first(similar_loss_lines.picking_id) # <- LOOK HERE
    return loss_picking
```

lmignon replied:
My bad 😕

jbaudoux replied:
Still `_find_loss_picking_moves_domain` needs to be renamed to `_find_loss_move_line_domain`

lmignon commented on stock_picking_operation_loss_quantity/models/stock_move_line.py:
```diff
@@ -0,0 +1,185 @@
+            )
+
+        # Search for an already existing LOSS picking for this quant
+        loss_picking = self._find_loss_picking()
```
This variable name is confusing `loss_picking` seems to refer to a picking but in fact it's a stock.move.line....

nicolas-delbovier-acsone replied:
It is a picking, see my comment above

lmignon reviewed (changes requested):
@nicolas-delbovier-acsone  Some little comments about methods and variable naming but it looks good 😏

lmignon reviewed (approved):
Code review + Functional tests

nicolas-delbovier-acsone committed:
[FIX] stock_picking_operation_loss_quantity: invalid loss sequence prefix

nicolas-delbovier-acsone committed:
[IMP] stock_picking_operation_loss_quantity: auto reallocate stock from another valid quants when a loss is declared

This allows the warehouse operator to potentially still complete the picking
with stock from another location/lot even after a loss declaration.

jbaudoux commented on stock_picking_operation_loss_quantity/models/stock_move.py:
```diff
@@ -0,0 +1,29 @@
+            )
+            > 0
+        ):
+            self.with_context(_loss_ignored_locations=ignored_location)._action_assign()
```
@nicolas-delbovier-acsone You need to give `force_qty` otherwise the reservation will be skipped if the line was partially processed as the move is in partially_available

nicolas-delbovier-acsone replied:
I've just tested this out but it breaks `test_loss_quantity_auto_reallocation` unit test with:
```python
>       self.assertEqual(new_line.reserved_uom_qty, 5.0)
E       AssertionError: 1.0 != 5.0
```

I updated the unit test to show that the code handles partial qty gracefully "as is". Can you have a look and tell me if the updated unit test covers what you had in mind?

jbaudoux replied:
Can you make a test where you set a quantity done on the second move line and then declare a loss on the first move line. For that use case, it requires the force_qty to equal to the move line quantity that was unlinked just before. However, I see this requirement is for odoo >= 17.0.
https://github.com/odoo/odoo/blob/17.0/addons/stock/models/stock_move.py#L1691

nicolas-delbovier-acsone replied:
Done! Is this new test what you expected? Note that I did not need to add the "force_qty"

jbaudoux replied:
Yes thanks for the test. Indeed you don't need force_qty in 16.0 but it will be required in next versions. To make it migration friendly, you could add it :)

nicolas-delbovier-acsone replied:
No I can not add it since it breaks the test as I mentioned above:

>I've just tested this out but it breaks test_loss_quantity_auto_reallocation unit test with:

nicolas-delbovier-acsone commented:
I had to change the logic for reallocation a bit in order to allow odoo to reserve stock from the same location but from another lot.

The logic changed from 
> Try to reserve stock from another **location**

to
>Try to reserve stock from another **quant**

jbaudoux commented on stock_picking_operation_loss_quantity/models/stock_move_line.py:
```diff
@@ -0,0 +1,199 @@
+            "product_uom_qty": unprocessed_qty,  # This is the "demand"
+            "product_uom": self.product_uom_id.id,
+            "location_id": self.location_id.id,
+            "location_dest_id": loss_pick_type.default_location_dest_id.id,
```
```suggestion
            "picking_type_id": loss_pick_type.id,
            "location_id": self.location_id.id,
            "location_dest_id": loss_pick_type.default_location_dest_id.id,
```

jbaudoux commented on stock_picking_operation_loss_quantity/models/stock_move_line.py:
```diff
@@ -0,0 +1,199 @@
+        return self._lose_quantity()
+
+    def _prepare_loss_move_vals(self, unprocessed_qty):
+        loss_pick_type = self.location_id.warehouse_id.loss_type_id
```
loss_pick_type should be an argument of the method

jbaudoux commented on stock_picking_operation_loss_quantity/models/stock_move_line.py:
```diff
@@ -0,0 +1,199 @@
+        new_loss_move_vals = self._prepare_loss_move_vals(unprocessed_qty)
+        new_loss_move = self.env["stock.move"].create(
+            {**new_loss_move_vals, "picking_id": loss_picking.id}
+        )
```
Set picking type reservation to "manual" and you should not bother of the picking anymore
```suggestion
        new_loss_move_vals = self._prepare_loss_move_vals(unprocessed_qty, loss_pick_type)
        new_loss_move = self.env["stock.move"].create(new_loss_move_vals)
```

jbaudoux commented on stock_picking_operation_loss_quantity/models/stock_move_line.py:
```diff
@@ -0,0 +1,199 @@
+            {**new_loss_move_vals, "picking_id": loss_picking.id}
+        )
+        new_loss_move._action_confirm()
+        loss_picking.loss_declaration_count += 1
```
the count should be on the move. The call to `_action_confirm` returns the new move after merge

jbaudoux commented on stock_picking_operation_loss_quantity/models/stock_move_line.py:
```diff
@@ -0,0 +1,199 @@
+        loss_picking = first(similar_loss_lines.picking_id)
+        return loss_picking
+
+    def _create_loss_move_line(self, unprocessed_qty: float):
```
Nice to have.
Move to `stock.quant` and call it `_lock_quant(picking_type, quantity)`
A base method that locks a quant by placing a reservation on it. There could be different types of locks using different picking types (loss, quality lock...)
That could even move to a separate base module `stock_quant_lock`

jbaudoux commented on stock_picking_operation_loss_quantity/models/stock_move_line.py:
```diff
@@ -0,0 +1,199 @@
+            self.location_id.warehouse_id.loss_auto_clear_threshold
+            and loss_picking.loss_declaration_count
+            >= self.location_id.warehouse_id.loss_auto_clear_threshold
+        ):
```
You could make this check before calling the method to give the right quantity to the created move. No need to create 2 times a move.
Would be more clear if the method was on the quant

jbaudoux commented on stock_picking_operation_loss_quantity/models/stock_move_line.py:
```diff
@@ -0,0 +1,199 @@
+                )
+                new_loss_move._action_confirm()
+
+        loss_picking.action_assign()
```
Bug !

You gather specific quant, create a move for the quantity but don't create the move line.
The reservation could lock another quant on the same location+product.

jbaudoux commented on stock_picking_operation_loss_quantity/models/stock_move_line.py:
```diff
@@ -0,0 +1,199 @@
+        whole quantity, do not allow to declare a loss).
+
+        Then, lock the quant that should be reserved by the loss picking and
+        create that loss picking.
```
Ah, it's written here :) There definitely should be somewhere a `quant._lock_quant(picking_type, quantity)`

jbaudoux commented on stock_picking_operation_loss_quantity/models/stock_move_line.py:
```diff
@@ -0,0 +1,199 @@
+        Then, lock the quant that should be reserved by the loss picking and
+        create that loss picking.
+        """
+        loss_lines = self.filtered(lambda line: line.progress != 100.0)
```
Overkill to depend on `stock_picking_progress` module just to compare if the done quantity == to do quantity.
Remove that dependency

jbaudoux commented on stock_picking_operation_loss_quantity/models/stock_move_line.py:
```diff
@@ -0,0 +1,199 @@
+        unprocessed_qty = self.reserved_uom_qty - self.qty_done
+        # Free the quantity that the operator was not able to do
+        self.reserved_uom_qty = self.qty_done
+        return unprocessed_qty
```
Convert to product uom

jbaudoux commented on stock_picking_operation_loss_quantity/models/stock_move_line.py:
```diff
@@ -0,0 +1,199 @@
+            "product_uom": self.product_uom_id.id,
+            "location_id": self.location_id.id,
+            "location_dest_id": loss_pick_type.default_location_dest_id.id,
+            "lot_ids": [Command.set(self.lot_id.ids)] if self.lot_id else False,
```
That's for serial numbers, not for reserving a specific lot

jbaudoux commented on stock_picking_operation_loss_quantity/models/stock_move_line.py:
```diff
@@ -0,0 +1,199 @@
+        new_loss_move = self.env["stock.move"].create(
+            {**new_loss_move_vals, "picking_id": loss_picking.id}
+        )
+        new_loss_move._action_confirm()
```
This may fail.
You don't control the reserved lot / owner.
You must create manually the move line.

jbaudoux commented on stock_picking_operation_loss_quantity/models/stock_move_line.py:
```diff
@@ -0,0 +1,199 @@
+
+            unprocessed_qty = line._unreserve_unprocessed_qty()
+            loss_picking = line._create_loss_move_line(unprocessed_qty)
+            loss_picking._schedule_loss_activity()
```
To rework. Something like:

```suggestion
            quants._lock_quants_for_loss()
            loss_pick_type = self.location_id.warehouse_id.loss_type_id

            unprocessed_qty = line._unreserve_unprocessed_qty()
            # todo : search for previous loss move, check count / threshold
            ....
            loss_pickings = self.env["stock.picking"]
            for quant in quants:
                lock_quantity = unprocessed_qty or quant.quantity
                move = quant._lock_quant(lock_quantity, loss_pick_type)
                loss_pickings |= move.picking_id
                # todo : decrease unprocessed_qty
                ...
            loss_pickings._schedule_loss_activity()
```
